src/coreclr/nativeaot/Runtime/thread.cpp

```diff
@@ -392,13 +392,26 @@ void Thread::Destroy()
 }
 
 #ifdef HOST_WASM
-extern OBJECTREF * GetShadowStackBottom();
-extern OBJECTREF * GetShadowStackTop();
+// TODO-LLVM-Bug: the shadow stack stuff needs to be moved to Thread for multi-threading to work properly.
+extern OBJECTREF* GetShadowStackBottom();
+extern OBJECTREF* GetShadowStackTop();
 
-void GcScanWasmShadowStack(ScanFunc * pfnEnumCallback, ScanContext * pvCallbackData)
+void Thread::GcScanWasmShadowStack(ScanFunc * pfnEnumCallback, ScanContext * pvCallbackData)
 {
     // Wasm does not permit iteration of stack frames so is uses a shadow stack instead
     EnumGcRefsInRegionConservatively(GetShadowStackBottom(), GetShadowStackTop(), pfnEnumCallback, pvCallbackData);
+
+    // TODO-LLVM-Upstream: unify this method with the general "GcScanRootsWorker" below.
+    for (GCFrameRegistration* pCurGCFrame = m_pGCFrameRegistrations; pCurGCFrame != NULL; pCurGCFrame = pCurGCFrame->m_pNext)
```

Comment:
The GC frames are usually only used for dynamic/reflection invokes (that need to preserve a dynamic number of GC pointers).